### PR TITLE
[4.x] Adds support for using recursion on arbitrary array data

### PR DESCRIPTION
**Pull request by JohnathonKoster**
This PR adds support for using `{{ *recursive children* }}` on arbitrary array data, and is a step towards generalizing the current `nav` tag fix in the codebase.

The following will now work with general array data or data from tags other than `nav`:

```antlers
{{ some_data }}
<ul>
    {{ children }}
        {{ title }}

        {{ if children }}
        <ul>
            {{ *recursive children* }}
        </ul>
        {{ /if }}
    {{ /children }}
</ul>
{{ /some_data }}
```

### CODE REVIEW

**Comment by jasonvarga** [comment]
> and is a step towards generalizing the current nav tag fix in the codebase.

What does this mean?

**Comment by JohnathonKoster** [comment]
> > and is a step towards generalizing the current nav tag fix in the codebase.
> 
> What does this mean?

It relates to this line:

https://github.com/statamic/cms/blob/4.x/src/View/Antlers/Language/Analyzers/RecursiveParentAnalyzer.php#L58

and our previous discussion here:

https://github.com/statamic/cms/pull/6968#discussion_r1008296645

File: src/View/Antlers/Language/Analyzers/RecursiveParentAnalyzer.php [hunk 1]
```diff
@@ -54,6 +54,10 @@ public static function associateRecursiveParent($nodes)
                             if (Str::contains($subNode->runtimeContent, $recursiveContent) && mb_substr_count($subNode->runtimeContent, '*recursive') == 1 && $node->getRootRef() == $subNode->getRootRef()) {
                                 $lastNode = $subNode;
 
+                                if (! ConditionPairAnalyzer::isConditionalStructure($subNode) && $subNode->isClosedBy != null && $subNode->name->name == $node->name->name) {
+                                    break;
+                                }
+
                                 // Ensure we stop searching once we reach the closest nav parent.
                                 if ($subNode->name->name == 'nav') {
```
Comment (jasonvarga) [inline src/View/Antlers/Language/Analyzers/RecursiveParentAnalyzer.php:62]: Does this PR allow us to remove this then?

Comment (JohnathonKoster) [reply src/View/Antlers/Language/Analyzers/RecursiveParentAnalyzer.php:62]: I'll test to see if we can safely do that :)

**Pull request merged**